elephant/gpfa.py

```diff
@@ -0,0 +1,402 @@
+                observation noise covariance
+
+    seqs_train: np.recarray
+        Data structure, whose n-th entry (corresponding to the n-th
+        experimental trial) has fields
+            * trialId: int
+                unique trial identifier
+            * T: int
+                number of timesteps
+            * y: ndarray of shape (#units, #bins)
+                neural data
+            * xsm: ndarray of shape (#latent_vars, #bins)
+                posterior mean of latent variables at each time bin
+            * Vsm: ndarray of shape (#latent_vars, #latent_vars, #bins)
+                posterior covariance between latent variables at each
+                timepoint
+            * VsmGP: ndarray of shape (#bins, #bins, #latent_vars)
+                posterior covariance over time for each latent variable
+
+    fit_info: dict
+        Information of the fitting process and the parameters used there:
+            * iteration_time: A list containing the runtime for each iteration
+                step in the EM algorithm.
+            * log_likelihood: float, maximized likelihood obtained in the
+                E-step of the EM algorithm.
+            * bin_size: int, Width of the bins.
+            * cvf: int, number for cross-validation folding
+                Default is 0 (no cross-validation).
+            * has_spikes_bool: Indicates if a neuron has any spikes across
+                trials.
+            * method: str, Method name.
+
+    Raises
+    ------
+    AssertionError
+        If `seqs` es empty.
+
+    """
+    assert len(seqs) > 0, "Got empty trials."
+
+    # Set cross-validation folds
+    num_trials = len(seqs)
+
+    test_mask = np.full(num_trials, False, dtype=bool)
+    train_mask = ~test_mask
+
+    tr = np.arange(num_trials, dtype=np.int)
+    train_trial_idx = tr[train_mask]
+    test_trial_idx = tr[test_mask]
+    seqs_train = seqs[train_trial_idx]
+    seqs_test = seqs[test_trial_idx]
+
+    # Remove inactive units based on training set
+    has_spikes_bool = (np.hstack(seqs_train['y']).mean(1) != 0)
+
+    for seq_train in seqs_train:
+        seq_train['y'] = seq_train['y'][has_spikes_bool, :]
+    for seq_test in seqs_test:
+        seq_test['y'] = seq_test['y'][has_spikes_bool, :]
+
+    # Check if training data covariance is full rank
+    y_all = np.hstack(seqs_train['y'])
+    y_dim = y_all.shape[0]
+
+    if np.linalg.matrix_rank(np.cov(y_all)) < y_dim:
+        errmesg = 'Observation covariance matrix is rank deficient.\n' \
+                  'Possible causes: ' \
+                  'repeated units, not enough observations.'
+        raise ValueError(errmesg)
+
+    if verbose:
+        print('Number of training trials: {}'.format(len(seqs_train)))
+        print('Number of test trials: {}'.format(len(seqs_test)))
+        print('Latent space dimensionality: {}'.format(x_dim))
+        print('Observation dimensionality: {}'.format(has_spikes_bool.sum()))
+
+    min_var_frac = 0.01
```

Comment:
This hard-coded number should rather be given as a keyword-argument (with a default of 0.01) of the function